broker/broker.go

```diff
@@ -383,24 +389,66 @@ func (b *S3Broker) Bind(
 	}
 
 	if _, err = b.user.Create(b.userName(bindingID), b.iamPath, iamTags); err != nil {
+		b.logger.Error("bind: error creating user", err, lager.Data{
+			instanceIDLogKey: instanceID,
+			bindingIDLogKey:  bindingID,
+			detailsLogKey:    details,
+			"user":           b.userName(bindingID),
+		})
 		return binding, err
 	}
+
 	defer func() {
+		// If the function returns an error, Bind did not complete and resources must be cleaned up.
 		if err != nil {
-			if policyARN != "" {
-				b.user.DeletePolicy(policyARN)
-			}
-			if accessKeyID != "" {
-				b.user.DeleteAccessKey(b.userName(bindingID), accessKeyID)
+			b.logger.Info("bind: defer: err was not nil on return; deleting user", lager.Data{
+				instanceIDLogKey: instanceID,
+				bindingIDLogKey:  bindingID,
+				detailsLogKey:    details,
+			})
+
+			// Careful: Do not shadow err, or future defers will not work.
+			if derr := b.user.Delete(b.userName(bindingID)); derr != nil {
+				b.logger.Error("bind: defer: error deleting user", derr, lager.Data{
```

Comment:
Do we do anything with `derr`? Should we?

Reply:
Besides logging it, no. Since this only gets called if we're already cleaning up from another error, the best I could think to do was log what happened. Unfortunately that leaves us in an inconsistent state.